### ISSUE TITLE
Consider backend-erased EagerTensor for dynamic tensor-network workloads

### ISSUE THREAD

**Issue by shinaoka** [tensor4all/tenferro-rs#867]
## Motivation

`EagerTensor<B: TensorBackend>` currently carries the backend as a Rust type parameter. This gives static dispatch, but it pushes backend choice into user-facing tensor types. For tensor-network libraries such as tensor4all, that is awkward: tensor-network objects are naturally dynamic containers, and making `TensorDynLen`, `TensorTrain`, `TreeTN`, etc. generic over `B` would cause broad API/type propagation.

In the tensor-network layer we would like `TensorDynLen` to become a thin wrapper around an eager payload plus index/layout metadata:

```rust
pub struct TensorDynLen {
    indices: Vec<DynIndex>,
    payload: Arc<TensorDynLenPayload>,
}

struct TensorDynLenPayload {
    eager: Arc<EagerTensor>,
    layout: TensorPayloadLayout,
}
```

That shape works much better if tenferro exposes a backend-erased `EagerTensor`.

## Proposal

Consider removing the backend type parameter from the public eager API and moving backend selection into the eager context:

```rust
pub struct EagerTensor {
    data: Arc<Tensor>,
    key: GlobalValKey<StdTensorOp>,
    grad_node: Option<Arc<GradNode<StdTensorOp>>>,
    requires_grad: bool,
    grad_slot: GradSlot,
    metadata_scopes: Vec<Arc<MetadataScope>>,
    ctx: Arc<EagerContext>,
}

pub struct EagerContext {
    id: ContextId,
    backend: Mutex<EagerBackend>,
}

pub enum EagerBackend {
    Cpu(CpuBackend),
    #[cfg(feature = "gpu")]
    Gpu(GpuBackend),
}
```

Operations would dispatch through the runtime backend enum:

```rust
impl EagerBackend {
    fn dot_general_with_conj(
        &mut self,
        lhs: &Tensor,
        rhs: &Tensor,
        config: &DotGeneralConfig,
        lhs_conj: bool,
        rhs_conj: bool,
    ) -> Result<Tensor> {
        match self {
            EagerBackend::Cpu(backend) => backend.dot_general_with_conj(lhs, rhs, config, lhs_conj, rhs_conj),
            EagerBackend::Gpu(backend) => backend.dot_general_with_conj(lhs, rhs, config, lhs_conj, rhs_conj),
        }
    }
}
```

## Design notes

- Prefer an enum-based facade over `dyn TensorBackend` unless the backend trait is intentionally made object-safe. The enum path is straightforward and keeps backend-specific capabilities explicit.
- Context mismatch checks should remain based on `EagerContext` identity, not just backend kind. This preserves gradient-slot, buffer-pool, and execution-session consistency.
- Static backend-specialized internals can remain private if useful; the public eager tensor type does not need to expose `B`.
- This would make tensor-network types runtime-backend aware without requiring generic propagation through `TensorDynLen`, `TensorTrain`, and `TreeTN`.

## Related tensor4all cleanup

This is motivated by a planned tensor4all `TensorDynLen` redesign. The current `TensorDynLen` stores both a `Storage` snapshot and an eager cache. AD information lives on tenferro `EagerTensor`, not in `Storage`, and contraction intermediates pay unnecessary materialization cost when converted back to `Storage`. A cleaner design is to make `TensorDynLen` primarily `EagerTensor + layout + indices`, with `Storage` demoted to an explicit/materialized snapshot compatibility layer.
